Lustre plugin for gufi_dir2index

bertschinger opened the pull request:
This change adds the ability to pass a plugin library to run user-specific code during gufi_dir2index.

The plugin included here gathers Lustre stripe information and FIDs for each file/directory and stores it into the sqlite3 database.

bertschinger committed:
set CMAKE_ENABLE_EXPORTS when building GUFI binaries

This corresponds to the `-export-dynamic` linker option.

This is needed for dynamically loading plugins that use sqlite3
functions.

GUFI binaries are built with the sqlite3 object file statically linked.
When a plugin is loaded that calls sqlite3 functons, it must refer to
the definitions of the sqlite3 symbols that live in the GUFI binary that
the plugin is loaded into.

The `-export-dynamic` linker option makes the GUFI binaries export these
symbols so that the plugin is able to resolve them correctly.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+    sqlite3_stmt *statement;
+    char *text = "CREATE TABLE lustre_summary (ost_index INTEGER PRIMARY KEY, num_files INTEGER);";
+
+    int res = sqlite3_prepare_v2(db, text, strlen(text) + 1, &statement, NULL);
+    if (res != SQLITE_OK) {
+        fprintf(stderr, "lustre plugin: error preparing sqlite3 statement: %d\n", res);
+        return NULL;
+    }
+
+    res = sqlite3_step(statement);
+    if (res != SQLITE_DONE) {
+        fprintf(stderr, "lustre plugin: error evaluting statement: %d\n", res);
+        return NULL;
+    }
+
+    sqlite3_finalize(statement);
```
`sqlite3_exec`

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    sqlite3_stmt *statement;
+    char *text = "CREATE TABLE lustre_summary (ost_index INTEGER PRIMARY KEY, num_files INTEGER);";
```
`static const char text[] ...` so you can do `sizeof(text) - 1`

calccrypto replied:
Would it make sense to add `IF NOT EXISTS`?

bertschinger replied:
> Would it make sense to add `IF NOT EXISTS`, or `DROP TABLE IF EXISTS lustre_summary;` before creating it?

Hmm... maybe? Isn't this always run on a fresh database though? Or is that not the case?

calccrypto replied:
I might have been remembering old behavior, when overwriting indexes, the databases were written to without being replaced by an empty template first.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+
+    for (uint32_t i = 0; i <= state->num_osts; i++) {
+        sqlite3_stmt *statement;
+        char *text = sqlite3_mprintf("INSERT INTO %s VALUES(%" PRIu32 ", %" PRIu64 ");",
+                "lustre_summary", i, state->stripe_count[i]);
+        int res = sqlite3_prepare_v2(db, text, -1, &statement, NULL);
+        if (res != SQLITE_OK) {
+            fprintf(stderr, "lustre plugin: error preparing sqlite3 statement: %d\n", res);
+            goto out;
+        }
+
+        res = sqlite3_step(statement);
+        if (res != SQLITE_DONE) {
+            fprintf(stderr, "lustre plugin: error evaluting statement: %d\n", res);
+        }
+
+        sqlite3_finalize(statement);
```
`sqlite3_exec`

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+void db_exit(sqlite3 *db, void *user_data) {
+    struct stripe_tracker *state = (struct stripe_tracker *) user_data;
+
+    for (uint32_t i = 0; i <= state->num_osts; i++) {
```
`<=`?

bertschinger replied:
This looks wrong but I think it's actually correct. I renamed the variable to `max_ost_index` to make it clearer what it represents. For example, if there are 4 OSTs, then `max_ost_index` would be 3, not 4.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+};
+
+static void *alloc_lum() {
+    size_t v1 = sizeof(struct lov_user_md_v1) +
+        LOV_MAX_STRIPE_COUNT * sizeof(struct lov_user_ost_data_v1);
+    size_t v3 = sizeof(struct lov_user_md_v3) +
+        LOV_MAX_STRIPE_COUNT * sizeof(struct lov_user_ost_data_v1);
```
`static const`

calccrypto commented on src/gufi_dir2index.c:
```diff
@@ -407,14 +434,19 @@ static void sub_help(void) {
 
     int rc = EXIT_SUCCESS;
 
+    if (load_plugin_library(&pa.in)) {
+        rc = EXIT_FAILURE;
+        goto cleanup;
+    }
+
```
load the library in `parse_cmd_line` and clean up in `input_fini`

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+#include <errno.h>
```
Add license

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+    uint32_t initial_size = 64;
+
+    struct stripe_tracker *new = malloc(sizeof *new);
+    if (!new) {
+        return NULL;
+    }
```
There's no way for this to ever be entered.

bertschinger replied:
IMO it's not reasonable to assume that every user will run GUFI in an environment with infallible memory allocations. We can't control if an admin turns off vm_overcommit on their server, for example.

I do agree that checking for `malloc()` failures is a waste of time (for this kind of program, at least, if not in general), but this isn't my preferred way to do it. Rather than pretend that `malloc()` is infallible (even when it is not always, due to factors outside of our control), I'd go with a genuinely infallible malloc wrapper, like:
```C
void *gufi_malloc(size_t size) {
    void *p = malloc(size);
    if (!p) {
        fprintf(stderr, "out of memory, aborting\n");
        exit(1);
    }
    return p;
}
```

I realize in many cases `malloc()` failing will soon terminate the process anyways as most pointers will be dereferenced at some point. While that has the same effect, it's harder to discover that OOM was the cause rather than any other kind of memory error. For example, in https://github.com/mar-file-system/GUFI/issues/165, we can postulate that a `malloc()` failure led to null pointer dereference -- but without more information, we don't know. Handling `malloc()` failures with an error message means we can identify such cases without needing to inspect a core file.

Anyways if you're not a fan we don't have to do that for GUFI -- but that's what I'd lean towards. Deliberately ignoring the error result of a function that can fail just feels like bad practice, and obscures the cause of a segfault.

calccrypto replied:
As far as I've seen, if you have virtual memory, you can allocate forever so long as you don't touch the allocations, so you'll never actually get a `NULL` (unless you allocate more total memory than you have in one go, such as `malloc((size_t) - 1)` which is not something any code should be doing), and thus you will never enter the error block. #165 probably returned a non-`NULL` pointer that did not have backing pages. Have you seen systems that will return `NULL` for reasonably sized allocations?

Additionally, assuming that you somehow do manage to get into the error, this prevents easy debugging because all of your allocations are in a centralized location, so you need a call stack to figure out which caller got the `NULL`, but instead of having an obvious invalid memory access that gdb stops on, you now have a clean exit, meaning the person doing the debugging has to put that much more effort into adding a breakpoint and continuing until memory runs out.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    new->stripe_count = calloc(initial_size, sizeof(*new->stripe_count));
+    if (!new->stripe_count) {
+        free(new);
+        return NULL;
+    }
```
There's no way for this to ever be entered.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+
+    for (uint32_t i = 0; i <= state->num_osts; i++) {
+        sqlite3_stmt *statement;
+        char *text = sqlite3_mprintf("INSERT INTO %s VALUES(%" PRIu32 ", %" PRIu64 ");",
+                "lustre_summary", i, state->stripe_count[i]);
```
Why is insertion part of exit and not process?

bertschinger replied:
Since this is inserting a summary, that is aggregate data based on every file seen in the directory, it seems to make more sense to do it once at the end rather than requiring every file to read-modify-write the database entry as its processed.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+        sqlite3_free(text);
+    }
+
+  out:
```
Label is only used once - just handle the error instead of jumping?

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+    struct stripe_tracker *tracker = (struct stripe_tracker *) user_data;
+    track_file_stripes(tracker, stripe_array, stripe_count);
+
+  out:
```
Label is only used once - just handle the error instead of jumping?

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+    size_t v3 = sizeof(struct lov_user_md_v3) +
+        LOV_MAX_STRIPE_COUNT * sizeof(struct lov_user_ost_data_v1);
+
+    return calloc(1, v1 > v3 ? v1 : v3);
```
Can this size be a compile time const?

calccrypto commented:
Can you also add a simple test plugin + tests?

calccrypto commented on src/gufi_dir2index.c:
```diff
@@ -156,6 +162,8 @@ static int process_nondir(struct work *entry, struct entry_data *ed, void *args)
         external_read_file(in, entry, process_external, nda->db);
     }
 
+    char *fullpath = strdup(entry->name);
```
only `strdup` if there's a plugin

bertschinger replied:
I handled this differently, by just removing the modification to `entry->name`. It seems like all tests still passed, and the only impact this would have (in less I missed something) would be an error message in `insertdbgo()` -- so maybe it's fine to just not trim `entry->name`?

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+    int rc = llapi_file_get_stripe(path, layout_info);
+
+    if (rc) {
+        printf("lustre plugin: error getting stripe info for %s: %s\n", path, strerror(errno));
```
`stderr`

calccrypto replied:
`errno` should be copied before being used

calccrypto replied:
While in this case you're fine, people tend to call other functions that they don't realize change `errno`, and end up getting prints that say that there are no errors even though they entered an error block. The way to prevent this is to copy `errno` so that side-effects don't mess up your error handling. I forget what I was doing, but I've seen this happen recently.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,262 @@
+#include "sqlite3.h"
+#include "lustre/lustreapi.h"
+
+char *my_basename(char *path) {
```
`static`

calccrypto commented on src/utils.c:
```diff
@@ -858,3 +859,28 @@ int gufi_dirfd(DIR *d) {
     return d_fd;
 }
+
+/*
+ * If a plugin library path is given in in->plugin_name,
+ * then load that library and add its plugin functions to in->plugin_ops.
+ *
+ * Returns 0 on success or 1 on failure.
+ */
+int load_plugin_library(struct input *in) {
+    if (!in->plugin_name.data) {
+        return 0;
+    }
+
+    void *lib = dlopen(in->plugin_name.data, RTLD_NOW);
+    if (!lib) {
+        fprintf(stderr, "Could not open plugin library: %s\n", dlerror());
+        return 1;
+    }
+
+    in->plugin_ops.db_init = dlsym(lib, "db_init");
+    in->plugin_ops.process_dir = dlsym(lib, "process_dir");
+    in->plugin_ops.process_file = dlsym(lib, "process_file");
+    in->plugin_ops.db_exit = dlsym(lib, "db_exit");
+
+    return 0;
+}
```
I'd say move this to its own file or into bf.c since that's the only place it is used.

calccrypto commented on src/utils.c:
```diff
@@ -858,3 +859,28 @@ int gufi_dirfd(DIR *d) {
+        return 0;
+    }
+
+    void *lib = dlopen(in->plugin_name.data, RTLD_NOW);
```
Does `dlclose` need to be called?

bertschinger replied:
Is unloading shared libraries normal practice? Valgrind, at least, doesn't seem to perceive it as a resource leak...

calccrypto replied:
I guess I have a habit of closing things I open.

bertschinger committed:
add plugin functionality

This enables the user to run custom code during database generation. The
user can populate the database with new tables, modify entries, etc.

bertschinger committed:
gufi_dir2index: allow running an optional user-specified plugin

The user can specify a plugin for gufi_dir2index so do custom
modifications to the databases as they are being created.

bertschinger committed:
add example Lustre plugin for adding lustre stripe info to the DBs

calccrypto commented on include/bf.h:
```diff
@@ -274,6 +297,12 @@ struct input {
    refstr_t swap_prefix;
+
+   /* A path to an optional plugin library */
+   refstr_t plugin_name;
+
+   /* If plugin_name is set, contains pointers to the plugin functions */
+   struct plugin_operations plugin_ops;
```
Maybe merge these into a single struct

bertschinger replied:
no longer relevant, since I changed the way it's loaded the plugin_name isn't stored in `struct input` now

bertschinger commented:
> Can you also add a simple test plugin + tests?

Yep, I'll work on that next.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,349 @@
+ *
+ *     $ cd contrib
+ *
+ *     $ gcc -g -O0 -c -fPIC -I../build/deps/sqlite3/include -I../include -I$LUSTRE_INCLUDE_DIR -I$LUSTRE_INCLUDE_DIR/uapi lustre_plugin.c
```
`../build/deps/sqlite3/include` -> `"${DEP_INSTALL_PREFIX}"`

calccrypto commented on src/bf.c:
```diff
@@ -137,6 +148,30 @@ void input_fini(struct input *in) {
+ *
+ * Returns 0 on success or 1 on failure.
+ */
+int load_plugin_library(struct input *in, char *plugin_name) {
```
`static`

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,349 @@
+IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+OF SUCH DAMAGE.
+*/
+
```
3 empty lines

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,349 @@
+ *
+ *     $ cd contrib
+ *
+ *     $ gcc -g -O0 -c -fPIC -I../build/deps/sqlite3/include -I../include -I$LUSTRE_INCLUDE_DIR -I$LUSTRE_INCLUDE_DIR/uapi lustre_plugin.c
+ *
+ *     $ gcc -shared -o liblustre_plugin.so lustre_plugin.o -L$LUSTRE_LIBRARY_DIR -llustreapi
```
Add to CMake

bertschinger committed:
gufi_dir2index: remove unnecessary manipulation of entry->name

`entry->name` has the full path, but was being overwritten with just the
relative path. This isn't actually necessary for any GUFI functionality.

Previously it was harmless, but it will be problematic for handing the
file to a plugin to process, since the plugin will need the full path.

So in preparation for hooking in the plugin, get rid of the modification
to entry->name.

bertschinger committed:
tests: add simple plugin for testing purposes

This adds a simple plugin for gufi_dir2index which will be used for a
regression test.

The plugin is not intended to be useful in the real world. It just does
enough to be useful for testing purposes.

calccrypto commented on contrib/lustre_plugin.c:
```diff
@@ -0,0 +1,351 @@
+    }
+
+    static const char text[] = "CREATE TABLE lustre_summary (ost_index INTEGER PRIMARY KEY, num_files INTEGER);";
+    char *error;
```
initialize

calccrypto commented on test/regression/gufi_plugin.sh.in:
```diff
@@ -0,0 +1,96 @@
+rm -rf "${SEARCH}"
+
+# generate the index
+"${GUFI_DIR2INDEX}" -U "@CMAKE_BINARY_DIR@/test/$PLUGIN_FILENAME" -x "${SRCDIR}" "${SEARCH}"
```
`${PLUGIN_FILENAME}`

calccrypto commented on test/regression/gufi_plugin.sh.in:
```diff
@@ -0,0 +1,96 @@
+rm -rf "${SEARCH}"
+
+# generate the index
+"${GUFI_DIR2INDEX}" -U "@CMAKE_BINARY_DIR@/test/$PLUGIN_FILENAME" -x "${SRCDIR}" "${SEARCH}"
```
use `run_sort` or `run_no_sort` to add command to test output

bertschinger replied:
Hmm, this seems somewhat problematic because the command is different depending on the environment, leading to test failures like this:
```
< $ gufi_dir2index -U libtest_plugin.so -x prefix search
---
> $ gufi_dir2index -U build/test/libtest_plugin.dylib -x prefix search
```

What do you think? How to best resolve this?

calccrypto commented on test/test_plugin.c:
```diff
@@ -0,0 +1,173 @@
+
+void process_file(char *path, sqlite3 *db, void *user_data) {
+    struct state *p = (struct state *) user_data;
+    p->n_files += 1;
```
maybe move this into `if (res == SQLITE_OK)`

bertschinger commented on test/test_plugin.c:
```diff
@@ -0,0 +1,173 @@
+void *db_init(sqlite3 *db) {
+    struct state *state = new_state();
+
+    char *text = "CREATE TABLE plugin_test_files (id INTEGER PRIMARY KEY AUTOINCREMENT, filename TEXT);"
```
How so? Are you referring to `char *` vs `static const char text[]`? If so I think the latter was done that way because at one point I was calling `strlen(text)`, which was changed to `sizeof(test)` -- but that doesn't appear relevant anymore since now I'm using `sqlite3_exec()` instead of `sqlite3_prepare()`. So at this point I'd say it'd make more sense to go back to `char *` in the lustre plugin.

calccrypto replied:
That works
